### PR TITLE
Preserve dep win rt

### PR DESCRIPTION
**Pull request by vitek-karas**
Both cases have type references to a different assembly and linker currently can't add a new assembly to the closure when analyzing `Type.GetType` - so `PreserveDepenency` attribute is necessary in this case.

This is necessary for correctness, but the downside is that the System.Runtime.WindowsRuntime.dll will pretty much always be included even for apps which are not WinRT apps. To fix this, the linker would need to be able to determine the value of `ApplicationModel.IsUap` and remove inaccessible branches in the code - since all of this code is hidden behind checks for the `IsUap`.

### CODE REVIEW

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/SynchronizationContext.Uap.cs [hunk 1]
```diff
@@ -48,6 +48,10 @@ public partial class SynchronizationContext
 
         private static Func<object, SynchronizationContext>? s_createSynchronizationContextDelegate;
 
+        // This is necessary because linker can't add new assemblies to the closure when recognizing Type.GetType
+        // so even though the GetType call below is analyzable, the PreserveDependency is still necessary to actually include
+        // the assembly in the trimmed closure.
```
Comment (stephentoub) [inline src/coreclr/src/System.Private.CoreLib/src/System/Threading/SynchronizationContext.Uap.cs]: Is this a moment-in-time problem? I actually prefer being explicit about the dependency with the PreserveDependency attribute, even if the linker can analyze the Type.GetType, but that aside, I'm not clear on why the linker would be able to add the assembly to the closure based on the attribute but not based on the Type.GetType call.  Is that just a current bug that'll be fixed?

**Review by stephentoub — APPROVED** [review]

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/SynchronizationContext.Uap.cs (hunk 1 above)
Comment (vitek-karas) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/SynchronizationContext.Uap.cs]: There is an issue for this: https://github.com/mono/linker/issues/943
And strictly speaking it's a bug - ideally linker would be ale to handle this just fine.
Realistically - it's really hard to fix - it would require changing how the linker works all up.

Ideally we would actually degrade the linker and simply warn every time there's an Assembly part of the type name specified. Currently it's basically up for chance if it will work or not (if the target assembly somehow got to the closure, it will work without warnings, if it didn't it will warn) - but I fear that's too much of a breaking change for the linker.

So in short - I think we will keep the attribute in CoreLib for the foreseeable future.

**Review by marek-safar — APPROVED** [review]
